File: java_hummus/src/main/java/com/github/klee0kai/hummus/threads/FutureHolder.java
package com.github.klee0kai.hummus.threads;

import com.github.klee0kai.hummus.AndroidDevKitLogs;

import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

public class FutureHolder<T> {

    private Future<T> future = null;

    public Future<T> set(Future<T> future) {
        return this.future = future;
    }

    public Future<T> get() {
        return future;
    }

    public boolean isInProcess() {
        return future != null && !future.isDone() && !future.isCancelled();
    }

    public void cancel() {
        if (future != null && !future.isDone())
            future.cancel(false);
    }


    public T tryAwait() {
        try {
            return future != null ? future.get() : null;
        } catch (Exception e) {
            AndroidDevKitLogs.w(e);
            return null;
        }
    }

    public T getResult() {
        try {
            return future != null && future.isDone() ? future.get(1, TimeUnit.MICROSECONDS) : null;
        } catch (Exception e) {
            AndroidDevKitLogs.w(e);
            return null;
        }
    }

    public T popResult() {
        try {
            T res = future != null && future.isDone() ? future.get(1, TimeUnit.MICROSECONDS) : null;
            if (res != null)
                future = null;
            return res;
        } catch (Exception e) {
            AndroidDevKitLogs.w(e);
            return null;
        }
    }


}

File: kdevkit/src/main/java/com/github/klee0kai/android_devkit/threads/FutureHolder.kt
package com.github.klee0kai.android_devkit.threads

import com.github.klee0kai.android_devkit.AndroidDevKitLogs.devKitLog
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

open class FutureHolder<T> {

    private var future: Future<T>? = null

    val isInProcess: Boolean
        get() = future?.isDone != true && future?.isCancelled != true


    fun set(future: Future<T>): Future<T> {
        return future.also { this.future = it }
    }

    fun get(): Future<T>? {
        return future
    }

    fun cancel() {
        if (future?.isDone != true) {
            future?.cancel(false)
        }
    }

    fun tryAwait(): T? = runCatching {
        future?.get()
    }.onFailure {
        devKitLog?.w(it)
    }.getOrNull()

    fun result(): T? = runCatching {
        if (future?.isDone == true) {
            future?.get(1, TimeUnit.MICROSECONDS)
        } else {
            null
        }
    }.onFailure {
        devKitLog?.w(it)
    }.getOrNull()


    fun popResult(): T? {
        return result()?.also {
            future = null
        }
    }
}